Add scripts related to ARX estimation of generalized plant

lucasp1701 opened the pull request:
Generalized plant is obtained from sysid experiments considering correctors to BPMs path.

lucasp1701 committed:
Add scripts related to ARX estimation of generalized plant

Generalized plant is obtained from sysid experiments considering
correctors to BPMs path.

Co-authored-by: Guilherme Ricioli <guilherme.riciolic@gmail.com>

guilhermerc commented on sysid/sysid_files.m:
```diff
@@ -0,0 +1,30 @@
+clc;clear;
+
+text = fileread("ps_names.txt");
```
If you're hardcoding filepaths, it's better to declare'em as variables on top of your script. This way, one can readily adjust these to its own local filepaths.

```
ps_names_fpath = "ps_names.txt";

text = fileread(ps_names_fpath);
```

guilhermerc commented on sysid/arquivos.m:
If I understood it right, this and `sysid_files.m` do the same thing. If so, please remove this one.

guilhermerc commented on sysid/modelo.m:
As we talked, please translate this to english.

guilhermerc commented on sysid/modelo.m:
```diff
@@ -0,0 +1,279 @@
+prbs_duration = A.mat_d.prbs_step_duration(1);
+moving_avg = double(A.mat_d.prbs_mov_avg_taps);
+
+%Remoção de transiente
```
How did you choose the first 19050 samples as the transient? By visual inspection, right? If so, could you state that in a comment? We should also check that acquisition has more than 19050 samples before trying to slice it. `assert` [1] does the trick. 

[1] https://www.mathworks.com/help/matlab/ref/assert.html

guilhermerc commented on sysid/modelo.m:
```diff
@@ -0,0 +1,279 @@
+orbx = double(A.mat_d.orbx);
+orby = double(A.mat_d.orby);
+sampling_frequency = A.mat_d.sampling_frequency;
+prbs_n = A.mat_d.prbs_lfsr_len(1);
```
I think `prbs_n` and `prbs_duration` are misleading names... I think of them being 'number of PRBS cycles' and 'length of the PRBS signal', which they're not. Can you use `prbs_lfsr_len`and `prbs_step_duration` instead?

guilhermerc replied:
And the same goes to `moving_avg`. The variable doesn't hold a moving average filter object. `prbs_mov_avg_taps` is more appropriate.

guilhermerc commented on sysid/modelo.m:
```diff
@@ -0,0 +1,279 @@
+%}
+
+%Aplicação no modelo ARX
+if plano=='X'
```
Since only one of the orbit planes are being used, what if you pick the appropriate one at the beginning of the function and only apply the filtering/averaging to it? This way you don't spend processing time with a signal that you won't use.

guilhermerc reviewed (changes requested):
General comments:

1. Use more descriptive script names.

Here are suggestions, but feel free to pick any other (descriptive) name:
sysid/modelo.m => sysid/plant_arx_fit.m
sysid/sysid_files.m => sysid/generate_plant_models.m

2. Add the following as a header to all of your files:
```
% Copyright (C) 2024 CNPEM (cnpem.br)
% Author: Lucas Pelike <lucas.pelike@lnls.br>
```

3. Check if everything that's commented is really useful.

guilhermerc commented on sysid/modelo.m:
```diff
@@ -0,0 +1,279 @@
+
+%Filtros (de repositório fofb/sysid)
+fir_remove_switching = dfilt.dffir(ones(1,2^moving_avg)/2^moving_avg);
+iir_lowpass = design(fdesign.lowpass(0.23, 0.25, 1, 80, 1), 'butter', 'MatchExactly', 'stopband');
```
What's the filter bandwidth of this filter?

guilhermerc commented on sysid/modelo.m:
```diff
@@ -0,0 +1,279 @@
+%Filtros (de repositório fofb/sysid)
+fir_remove_switching = dfilt.dffir(ones(1,2^moving_avg)/2^moving_avg);
+iir_lowpass = design(fdesign.lowpass(0.23, 0.25, 1, 80, 1), 'butter', 'MatchExactly', 'stopband');
+currdata = filter(fir_remove_switching, currdata);
```
Since we're fitting an open-loop model, current doesn't have switching noise, only BPMs.

guilhermerc commented on sysid/modelo.m:
```diff
@@ -0,0 +1,279 @@
+function [sys,fit] = modelo(arquivo, parametros_arx, plano, aquisicao)
+
```
Please, document the function here. You should describe its purpose and its input/output arguments. Use [ofbmdl documentation](https://github.com/lnls-dig/fofb/blob/572a5f8400df62de40aff22593b32f03b9775617/ofbmdl.m#L2-L33) as a reference.

guilhermerc commented on sysid/modelo.m:
```diff
@@ -0,0 +1,279 @@
+
+%Redução do tempo de aquisição (opcional)
+T_aq=single(h2)/sampling_frequency;
+if exist('aquisicao', 'var')
```
I think it'd be better if this argument was the number of samples to take into account for the model fitting. We gain nothing but some inaccuracy by abstracting it to time...

guilhermerc commented on sysid/prbs_acq_h5tomat.m:
```diff
@@ -0,0 +1,60 @@
+
+clc;clear;
+ps_names_fpath = "ps_names.txt";
+bpm_fpath = "relation_corrector_to_bpm_idx_max_response.txt";
```
bad name: `bpm_fpath`

guilhermerc commented on sysid/prbs_acq_h5tomat.m:
```diff
@@ -0,0 +1,60 @@
+clc;clear;
+ps_names_fpath = "ps_names.txt";
+bpm_fpath = "relation_corrector_to_bpm_idx_max_response.txt";
+text = fileread(ps_names_fpath);
```
bad names: `text`, `text_bpm`

guilhermerc commented on sysid/prbs_acq_h5tomat.m:
```diff
@@ -0,0 +1,60 @@
+text = fileread(ps_names_fpath);
+text_bpm = fileread(bpm_fpath);
+expr = '[^\n]*[^\n]';
+matches = regexp(text,expr,'match')';
```
bad names: `matches`, `matches_bpm`

guilhermerc commented on sysid/prbs_acq_h5tomat.m:
```diff
@@ -0,0 +1,60 @@
+    tic;
+    file_aux=matches{i};
+    file = [file_aux '.h5'];
+    if (i~=1)&&(i~=80)&&(i~=81)&&(i~=160) %excluded correctors 1 80 81 160
```
no identation

guilhermerc commented on sysid/prbs_acq_h5tomat.m:
```diff
@@ -0,0 +1,60 @@
+for i=1:ncorr
+    fprintf("Corrector %d\n", i);
+    tic;
+    file_aux=matches{i};
```
bad names: `file`, `file_aux`

guilhermerc commented on sysid/prbs_acq_h5tomat.m:
```diff
@@ -0,0 +1,60 @@
+    file = [file_aux '.h5'];
+    if (i~=1)&&(i~=80)&&(i~=81)&&(i~=160) %excluded correctors 1 80 81 160
+    acq_nrpoints_before = h5read(file,'/params/acq_nrpoints_before');
+    prbs_signal = double(h5read(file,'/data/prbs_data/0'));
```
nit: for consistency, rename `prbs_signal` to `prbs_data`

guilhermerc commented on sysid/prbs_acq_h5tomat.m:
```diff
@@ -0,0 +1,60 @@
+    if (i~=1)&&(i~=80)&&(i~=81)&&(i~=160) %excluded correctors 1 80 81 160
+    acq_nrpoints_before = h5read(file,'/params/acq_nrpoints_before');
+    prbs_signal = double(h5read(file,'/data/prbs_data/0'));
+    prbs_signal = prbs_signal(acq_nrpoints_before:end);
```
As we talked, the slicing should be `acq_nrpoints_before+1:end`. Fix this everywhere its needed.

guilhermerc commented on sysid/prbs_acq_h5tomat.m:
This script is only applicable for open-loop acquisitions, so `sysid/prbs_ol_acq_h5_to_mat.m` is more appropriate.

guilhermerc replied:
Also, acquisitions are assumed to be in the folder as this script. A variable specifying their base path should be used instead.

guilhermerc replied:
Let's use this script for generating data objects that can be inputted into both generalized plant fitting and experimental sigma scripts.

guilhermerc replied:
@lucasp1701

guilhermerc commented on sysid/prbs_acq_h5tomat.m:
```diff
@@ -0,0 +1,60 @@
+    acq_nrpoints_before = h5read(file,'/params/acq_nrpoints_before');
+    prbs_signal = double(h5read(file,'/data/prbs_data/0'));
+    prbs_signal = prbs_signal(acq_nrpoints_before:end);
+    corrector_label = file(6:7);
```
`corr_type_label` is a better name

guilhermerc commented on sysid/prbs_acq_h5tomat.m:
```diff
@@ -0,0 +1,60 @@
+    orbx_aux = h5read(file,'/data/orbx');
+    if bpm_index<=floor(ncorr/2)
+        orbx_aux = orbx_aux(bpm_index,:);
+        data.orbx = orbx_aux(acq_nrpoints_before:end)';
```
this can be outside if-else statement

guilhermerc commented on sysid/prbs_acq_h5tomat.m:
```diff
@@ -0,0 +1,60 @@
+        orbx_aux = orbx_aux(bpm_index,:);
+        data.orbx = orbx_aux(acq_nrpoints_before:end)';
+        orby_aux = orby_aux(bpm_index,:);
+        data.orby = orby_aux(acq_nrpoints_before:end)';
```
this can be outside if-else statement

guilhermerc commented on sysid/prbs_acq_h5tomat.m:
```diff
@@ -0,0 +1,60 @@
+    data.input_signal = prbs_signal;
+    data.prbs_lfsr_len = h5read(file,'/data/prbs_lfsr_len');
+    data.prbs_step_duration = h5read(file,'/data/prbs_step_duration');
+    plane_label=file(end-3:end-3);
```
`plane_label` isn't being used

guilhermerc commented on sysid/prbs_acq_h5tomat.m:
```diff
@@ -0,0 +1,60 @@
+        data.orby = orby_aux(acq_nrpoints_before:end)';
+    else
+        orbx_aux = orbx_aux(bpm_index-80,:);
+        data.orbx = orbx_aux(acq_nrpoints_before:end)';
```
this can be outside if-else statement

guilhermerc commented on sysid/prbs_acq_h5tomat.m:
```diff
@@ -0,0 +1,60 @@
+    if strcmp(corrector_label,'C3')
+        prbs_signal = prbs_signal./2;
+    end
+    data.input_signal = prbs_signal;
```
nit: for consistency, rename `input_signal` to `prbs_data`

guilhermerc commented on sysid/prbs_acq_h5tomat.m:
```diff
@@ -0,0 +1,60 @@
+matches_bpm = regexp(text_bpm,expr,'match')';
+ncorr = size(matches,1);
+
+sys=cell(ncorr, 1);
```
`sys` isn't used

guilhermerc committed:
prbs_ol_acq_h5_to_mat.m: add script

Add script for converting HDF5 PRBS open-loop acquisition files into
MATLAB struct objects.
